jerry-core/ecma/operations/ecma-function-object.c

```diff
@@ -453,12 +453,164 @@ ecma_op_function_has_construct_flag (const ecma_value_t *arguments_list_p) /**<
 {
 #ifndef CONFIG_DISABLE_ES2015_CLASS
   return (((uintptr_t) arguments_list_p) & ECMA_CLASS_CONSTRUCT_FLAG);
-#else
+#else /* CONFIG_DISABLE_ES2015_CLASS */
   JERRY_UNUSED (arguments_list_p);
   return false;
 #endif /* !CONFIG_DISABLE_ES2015_CLASS */
 } /* ecma_op_function_has_construct_flag */
 
+#ifndef CONFIG_DISABLE_ES2015
+/**
+ * Returns the closest declarative lexical enviroment to the super object bound lexical enviroment.
+ *
+ * @return the found lexical enviroment
+ */
+static ecma_object_t *
+ecma_op_find_super_declerative_lex_env (ecma_object_t *lex_env_p) /**< starting lexical enviroment */
+{
+  JERRY_ASSERT (lex_env_p);
+  JERRY_ASSERT (ecma_op_resolve_super_reference_value (lex_env_p));
+  JERRY_ASSERT (ecma_get_lex_env_type (lex_env_p) != ECMA_LEXICAL_ENVIRONMENT_SUPER_OBJECT_BOUND);
+
+  while (true)
+  {
+    ecma_object_t *lex_env_outer_p = ecma_get_lex_env_outer_reference (lex_env_p);
+
+    if (ecma_get_lex_env_type (lex_env_outer_p) == ECMA_LEXICAL_ENVIRONMENT_SUPER_OBJECT_BOUND)
+    {
+      JERRY_ASSERT (lex_env_p);
```

Comment:
Moving the assert is OK for me.

The are seems similar to each other, but actually they do two different things.

`ecma_op_resolve_super_reference_value` finds the closest `SUPER_OBJECT_BOUND` lexical environment and returns with it's binding object to get the current context super reference.

`ecma_op_find_super_declerative_lex_env` is a bit more specific.
If a class heritage is present there **must** be already a  `SUPER_OBJECT_BOUND` lexical environment in the lex-env chain. Since `SUPER_OBJECT_BOUND` lex-envs wokrs similar to `THIS_OBJECT_BOUND` lex-envs and the required `CLASS_THIS_BINDING` internal property cannot be part of a `SUPER_OBJECT_BOUND` lexical environment (if it could there would be an issue while creating two ore more instances from the same class). So each class constructor put a new `DECLARATIVE` lexical environment on the top of their scope where we can _hide_ this internal property. Furthermore, an `eval ('super()')` can also add a new lexical environment to the chain as so on...

So both functions scan the lex-env chain, but since `ecma_op_resolve_super_reference_value` finds the lex-env binding object,  `ecma_op_find_super_declerative_lex_env` finds that specific `DECLARATIVE` lex-env on the top of `SUPER_OBJECT_BOUND` where this hidden internal property belongs to.